watched_tasks: maintain a list of spawned async tasks and propagate errors

hnez opened the pull request:
This PR is based on an idea from [a discussion](https://github.com/linux-automation/tacd/pull/43#pullrequestreview-1650219550) in PR #43, where it was noticed that errors from tasks and threads propagate nowhere.

This PR fixed that by assembling a list of threads and tasks that should __not__ complete before the end of the program and propagating all errors from these threads and tasks to `main()` and beyond.

This touches a lot of places, as we spawn quite a lot of async_std tasks (the `tacd` logs now say "Spawned 58 tasks tasks and 5 threads").

This PR allows us to add better error handling for tasks and threads (we can for example use the `?` operator to propagate errors out and do not have to resort to `.unwrap()` to at least crash the `tacd` from within a task on error). Adding this better error handling is out of scope for this PR, as it is already large enough.

hnez commented:
The failed `cargo deny` run should be fixed by the dependency updates in #42 and has nothing to do with this PR.

KarlK90 commented on src/adc/iio/hardware.rs:
```diff
@@ -342,102 +340,97 @@ impl IioThread {
+                    (thread, channels, buf)
+                }
+                Err(e) => {
+                    // Can not fail in practice as the queue is known to be empty
+                    // at this point.
+                    thread_res_tx.try_send(Err(e)).unwrap();
```
Can you put this into the expect call? In the future we want to deny unwraps completely.

hnez replied:
Sure. I hope to get rid of these panics rather sooner than later when #48 is merged, but until then we can have `.expect()` here.

KarlK90 replied:
Nice, lgtm!

KarlK90 commented on src/adc/iio/hardware.rs:
```diff
@@ -342,102 +340,97 @@ impl IioThread {
+                    // If the queue was already used just print an error instead.
+                    if let Some((_, tx)) = signal_ready.take() {
                         // Can not fail in practice as the queue is only .take()n
                         // once and thus known to be empty.
-                        tx.try_send(Ok(content)).unwrap();
+                        tx.try_send(Err(Error::new(e))).unwrap();
```
Can you put this into the expect call? In the future we want to deny unwraps completely.

hnez replied:
Sure. I hope to get rid of these panics rather sooner than later when #48 is merged, but until then we can have `.expect()` here.

KarlK90 replied:
Nice, lgtm!

KarlK90 commented on src/adc/iio/hardware.rs:
```diff
@@ -342,102 +340,97 @@ impl IioThread {
+                // Now that we know that the ADC actually works and we have
+                // initial values: return a handle to it.
+                if let Some((content, tx)) = signal_ready.take() {
+                    // Can not fail in practice as the queue is only .take()n
+                    // once and thus known to be empty.
+                    tx.try_send(Ok(content)).unwrap();
```
Can you put this into the expect call? In the future we want to deny unwraps completely.

hnez replied:
Sure. I hope to get rid of these panics rather sooner than later when #48 is merged, but until then we can have `.expect()` here.

KarlK90 replied:
Nice, lgtm!

KarlK90 reviewed (changes requested):
Nice! I like the changes and implementation. I thought whether or not the async runtime provides mechanisms to query running tasks and see if anything failed like [`tokio-console`](https://docs.rs/tokio-console/latest/tokio_console/), but this might be messy (or not even possible). Anyway, there are some minor things that should be corrected but apart from that LGTM.

hnez commented on src/watched_tasks.rs:
```diff
@@ -0,0 +1,282 @@
+    /// Future will return the Result of said task.
+    /// The WatchedTasks Future should be .awaited at the end of main() so
+    /// that the program ends if any of the watched tasks ends.
+    pub fn spawn_task<S, F>(&mut self, name: S, future: F)
+    where
+        S: Into<String>,
+        F: Future<Output = TaskResult> + Send + 'static,
+    {
+        let task = task::Builder::new()
+            .name(name.into())
+            .spawn(future)
+            .expect("cannot spawn task");
+
+        self.tasks.push(task);
+    }
```
We previously used `async_std:task::spawn()` which did [`.expect()`](https://docs.rs/async-std/latest/src/async_std/task/spawn.rs.html#31) in the background.
But I will have a look …

KarlK90 replied:
Sure, but as we implement our own api now it would be good to be consistent for both spawning `tasks` and `threads`.

hnez replied:
I've implemented this in a separate commit so it can be reviewed separately, as it touches a whole lot of files.

KarlK90 replied:
Thanks, lgtm.

KarlK90 commented on src/watched_tasks.rs:
```diff
@@ -0,0 +1,282 @@
+            )));
+        }
+
+        let handle = self.handle.take().expect("Failed to get thread handle");
```
```suggestion
        let handle = self
            .handle
            .take()
            .with_context(|| "Failed to get thread handle")?;
```

hnez replied:
Wouldn't work here as the `Future:poll()` implementation has to return a `Poll<>`, but I came up with another solution:

https://github.com/linux-automation/tacd/blob/69d52f700eb62493607edb6786486a6796b2d13a/src/watched_tasks.rs#L99-L108

KarlK90 replied:
Yes, I've seen this change. I'm not that familiar with async Rust (yet) but from my understanding it seems sound.

hnez commented:
While addressing your comments I have also changed the way `ThreadHandle` notifies a task about the completion of a thread.
Using a `async_std::channel` there was somewhat of a hack so I do not have to work with `Future` and `Waker` directly, because that sounded somewhat scary.
As it turns out working with `Waker`s is not scary at all, so we can just use one here:

https://github.com/linux-automation/tacd/blob/69d52f700eb62493607edb6786486a6796b2d13a/src/watched_tasks.rs#L29-L54

This does however mean that you have to re-review that part.

hnez commented:
The failed `cargo clippy` run should be fixed by #49 and has nothing to do with this PR.

KarlK90 reviewed (approved):
This looks good to me now and will definitely improve the error handling story in the code base. Feel free to merge.

jluebbe commented:
Adding a custom wrapper around standard APIs seems very invasive to me. Is this really such an unusual problem that there is no idiomatic way?

hnez commented:
Hi @jluebbe

> no idiomatic way

Not in `async_std` itself, so we would have to use a crate for it. I had a look around and found [`async_nursery`](https://docs.rs/async_nursery/) which may be coerced into doing what we we want (minus the thread handling), but I would prefer not to use it because it does not look too mature (and does not support threads in addition to tasks).

I quite like the tailor-made solution implemented in this PR and think 206 not-that-dense lines of code are an acceptable complexity compared to adding a new dependency to handle this for us.

I would prefer merging this as-is rather than adding a new dependency.

hnez commented:
The recently merged pull requests #33 and #34 made some changes to this PR necessary, as both of them spawn tasks.
Please re-review affected files.

hnez commented:
I've rebased this PR on top of recent `main` and think it should be the next one we merge.

KarlK90 commented:
I remember that @jluebbe wanted some written commentary why we choose to implement a bespoke solution and not use `async_nursery` or tokio's [JoinSet](https://docs.rs/tokio/1.33.0/tokio/task/struct.JoinSet.html). Could you add this to `watched_tasks.rs`? After that I'm happy to approve this.

hnez commented:
I've added a comment …

https://github.com/linux-automation/tacd/blob/6c7a2cd7fb57b626fb3f8abeac0bf367493eeb81/src/watched_tasks.rs#L11-L29

… that talks about why we went with our own task spawning wrapper instead of using something off-the-shelf.

KarlK90 reviewed (approved):
@hnez The comment summarizes the discussion very well. You have my approval - feel free to merge.
